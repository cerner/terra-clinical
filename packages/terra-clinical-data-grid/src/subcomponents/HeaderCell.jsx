import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import IconCaretUp from 'terra-icon/lib/icon/IconCaretUp';
import IconCaretDown from 'terra-icon/lib/icon/IconCaretDown';

import Cell from './Cell';
import ResizeHandle from './ResizeHandle';

import styles from './HeaderCell.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * String identifier of the column in which the HeaderCell will be rendered.
   */
  columnId: PropTypes.string.isRequired,
  /**
   * String-formatted width that the HeaderCell should be rendered as. Any valid css width value is supported (i.e. 200px, 3rem).
   */
  width: PropTypes.string.isRequired,
  /**
   * String text that will be rendered within the HeaderCell's default text region.
   */
  text: PropTypes.string,
  /**
   * String indicating the sort direction of the column. If provided, a sorting indicator will be
   * rendered within in the header. However, if custom children are also provided, the sorting indicator will
   * not be rendered.
   */
  sortIndicator: PropTypes.oneOf(['ascending', 'descending']),
  /**
   * Boolean indicating whether or not the HeaderCell is selectable.
   */
  isSelectable: PropTypes.bool,
  /**
   * Function that will be called when the HeaderCell is selected. The `isSelectable` prop must be true for
   * this function to be called.
   */
  onSelect: PropTypes.func,
  /**
   * Boolean indicating whether or not the HeaderCell is resizable. If true, a ResizeHandle will be rendered
   * over the HeaderCell's contents.
   */
  isResizable: PropTypes.bool,
  /**
   * Function that will be called upon the release of the ResizeHandle. The `isResizable` prop must be be true
   * for this function to be called. Parameters: `onResizeEnd(columnId, resizeDelta)`
   */
  onResizeEnd: PropTypes.func,
  /**
   * Content that will rendered within the HeaderCell. If provided, the `text` and `sortIndicator` props will be ignored.
   */
  children: PropTypes.node,
  /**
   * Function that will be called with a ref to the HeaderCell's selectable element. Parameters: `selectableRefCallback(selectableRef)`
   */
  selectableRefCallback: PropTypes.func,
};

class HeaderCell extends React.Component {
  constructor(props) {
    super(props);

    this.handleCellSelect = this.handleCellSelect.bind(this);
  }

  handleCellSelect(sectionId, rowId, columnId) {
    const { onSelect } = this.props;

    if (onSelect) {
      onSelect(columnId);
    }
  }

  render() {
    const {
      columnId,
      text,
      sortIndicator,
      isSelectable,
      isResizable,
      width,
      children,
      onResizeEnd,
      selectableRefCallback,
    } = this.props;

    let content = children;
    if (!content && (text || sortIndicator)) {
      let sortIndicatorComponent;
      if (sortIndicator) {
        sortIndicatorComponent = (
          <div className={cx('sort-indicator')}>
            {sortIndicator === 'ascending' ? <IconCaretUp /> : null}
            {sortIndicator === 'descending' ? <IconCaretDown /> : null}
          </div>
        );
      }

      content = (
        <div className={cx('default-layout')}>
          <div className={cx('text')}>
            {text}
          </div>
          {sortIndicatorComponent}
        </div>
      );
    }

    return (
      <Cell
        className={cx('header-cell')}
        sectionId=""
        rowId=""
        columnId={columnId}
        width={width}
        isSelectable={isSelectable}
        onSelect={this.handleCellSelect}
        selectableRefCallback={selectableRefCallback}
        aria-hidden={!content || undefined} // Using '|| undefined' to unset the attribute instead of setting it to false.
      >
        {content}
        {isResizable ? <ResizeHandle id={columnId} onResizeStop={onResizeEnd} /> : null }
      </Cell>
    );
  }
}

HeaderCell.propTypes = propTypes;

export default HeaderCell;
