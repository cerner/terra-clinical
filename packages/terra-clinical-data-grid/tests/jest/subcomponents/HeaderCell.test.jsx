import React from 'react';
import HeaderCell from '../../../src/subcomponents/HeaderCell';

describe('HeaderCell Snapshot Tests', () => {
  it('should render a HeaderCell with minimal props', () => {
    const cell = shallow((
      <HeaderCell
        columnId="column-test"
        width="123px"
      />
    ));

    expect(cell).toMatchSnapshot();
  });

  it('should render a Cell with all provided props for default rendering', () => {
    const cell = shallow((
      <HeaderCell
        columnId="column-test"
        width="123px"
        text="Test HeaderCell"
        sortIndicator="ascending"
        isSelectable
        onSelect={() => {}}
        selectableRefCallback={() => {}}
        isResizable
        onResizeEnd={() => {}}
      />
    ));

    expect(cell).toMatchSnapshot();
  });

  it('should render a Cell with custom contents if provided', () => {
    const cell = shallow((
      <HeaderCell
        columnId="column-test"
        width="123px"
        text="Test HeaderCell (this should not be rendered)"
        sortIndicator="ascending"
        isSelectable
        onSelect={() => {}}
        selectableRefCallback={() => {}}
        isResizable
        onResizeEnd={() => {}}
      >
        <div>Custom HeaderCell Content</div>
      </HeaderCell>
    ));

    expect(cell).toMatchSnapshot();
  });
});
