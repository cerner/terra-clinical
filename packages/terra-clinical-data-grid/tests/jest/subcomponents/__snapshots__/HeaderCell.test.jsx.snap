// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderCell Snapshot Tests should render a Cell with all provided props for default rendering 1`] = `
<Cell
  className="header-cell"
  columnId="column-test"
  isSelectable={true}
  onSelect={[Function]}
  rowId=""
  sectionId=""
  selectableRefCallback={[Function]}
  width="123px"
>
  <div
    className="default-layout"
  >
    <div
      className="text"
    >
      Test HeaderCell
    </div>
    <div
      className="sort-indicator"
    >
      <IconCaretUp
        className=""
        isBidi={true}
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
  </div>
  <ResizeHandle
    id="column-test"
    onResizeStop={[Function]}
  />
</Cell>
`;

exports[`HeaderCell Snapshot Tests should render a Cell with custom contents if provided 1`] = `
<Cell
  className="header-cell"
  columnId="column-test"
  isSelectable={true}
  onSelect={[Function]}
  rowId=""
  sectionId=""
  selectableRefCallback={[Function]}
  width="123px"
>
  <div>
    Custom HeaderCell Content
  </div>
  <ResizeHandle
    id="column-test"
    onResizeStop={[Function]}
  />
</Cell>
`;

exports[`HeaderCell Snapshot Tests should render a HeaderCell with minimal props 1`] = `
<Cell
  aria-hidden={true}
  className="header-cell"
  columnId="column-test"
  onSelect={[Function]}
  rowId=""
  sectionId=""
  width="123px"
/>
`;
